lib/services/access_request.go

```diff
@@ -1384,20 +1414,79 @@ func (m *RequestValidator) truncateTTL(ctx context.Context, identity tlsca.Ident
 	return ttl, nil
 }
 
+// getResourceViewingRoles gets the subset of the user's roles that could be used
+// to view resources (i.e. base roles + search as roles).
+func (m *RequestValidator) getResourceViewingRoles(ctx context.Context) ([]string, error) {
+	// No need to filter by resource IDs as that will be done later.
+	searchAsRoles, err := m.applicableSearchAsRoles(ctx, nil, "")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return apiutils.Deduplicate(slices.Concat(searchAsRoles, m.userState.GetRoles())), nil
+}
+
 // GetRequestableRoles gets the list of all existent roles which the user is
 // able to request.  This operation is expensive since it loads all existent
 // roles in order to determine the role list.  Prefer calling CanRequestRole
-// when checking against a known role list.
-func (m *RequestValidator) GetRequestableRoles() ([]string, error) {
-	allRoles, err := m.getter.GetRoles(context.TODO())
+// when checking against a known role list. If resource IDs or a login hint
+// are provided, roles will be filtered to only include those that would
+// allow access to the given resource with the given login.
+func (m *RequestValidator) GetRequestableRoles(ctx context.Context, resourceIDs []types.ResourceID, loginHint string) ([]string, error) {
+	allRoles, err := m.getter.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resources, err := m.getUnderlyingResourcesByResourceIDs(ctx, resourceIDs)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	cluster, err := m.getter.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	roles, err := m.getResourceViewingRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	accessChecker, err := NewAccessChecker(&AccessInfo{
+		Roles:    roles,
+		Traits:   m.userState.GetTraits(),
+		Username: m.userState.GetName(),
+	}, cluster.GetClusterName(), m.getter)
```

Comment:
This doesn't copy the `AllowedResourceIDs` from the user's original `AccessInfo`, it's important to copy that in case the user has already assumed a resource access request

Reply:
Hmm. I'm looking at [`accessChecker.CheckAccess`](https://github.com/gravitational/teleport/blob/1207c20f2d15e48f38c4bb79004c663bf852a7bb/lib/services/access_checker.go#L436), and it looks to me like even if `AllowedResourceIDs` set, `CheckAccess` will fail if the user doesn't also have the right roles, so I'm not sure what `AllowedResourceIDs` adds. Am I just misunderstanding the code?

Reply:
Right, but we also *want* CheckAccess to fail if the user has the right roles but **doesn't have** the right `AllowedResourcesIDs`.

In general once you assume a resource access request we only let you list resources specifically allowed by that resource access request. It would prevent discovering resources as if you had a transitive resource access request like role A has `search_as_roles: [B]` which has `search_as_roles: [C]`, which is an unlikely configuration because you will never be able to actually get role C, but if the user gets role B with a resource access request they shouldn't be able to list all resources allowed by role C.

AllowedResourceIDs is an important part of the AccessInfo, why leave it out without a good reason?

Reply:
Gotcha, I misunderstood the purpose of `AllowedResourceIDs`. I've added it in.